### PR TITLE
Searching on Binned Edges

### LINKED ISSUE #72: Closest Edge
It can occur that the closest edge to return is actually not the edge that is returned. picture the following:

     ______________________________________________________
                        \         a 
                         \
                          \ c
                           \
                            \__________
                                  b
                             
                             
                             
                                  X

The problem here is that we use a radius check to determine which edges are really reachable from a given spot. Clearly the correct edge to return  here would be `b` but the edge that gets returned is `a` simply because `a` has a much larger length and so its radius of influence encompasses `X` where `b`'s does not.

Clearly the radius approach would work well in a very dense urban area or even a place with homogeneously-lengthed edges, but sadly as is always the case, the real world is never like this. So the fix...

Ah but wait, you might be thinking.. we cant talk about a fix until we understand this 'radius of influence' non-sense. Ok fine.. The simple answer is we do this so we can quickly reject a large portion of edges that cant be all that close to the input. And in truth it works, except for when the best answer is reasonably far away from the input.

So yeah whats the fix?

* we could consider more edges by loosening the radius, this will make everything slower and still miss some cases
* we could look at the number of edges or nodes in a tile and decide to search them all if the tile is sparse, shouldnt hurt performance if we do it right but you could still get the wrong answer
* we could set a mask on every edge and only consider edges whose mask is close to the input location

lets exapnd on that last one (beause its a huge feature request...) link to feature request issue is forthcoming...

### PR DESCRIPTION
**Pull request by kevinkreiser**
closes #72 #62 #52 

So there were many problems with the ~~initial~~ second implementation of edge correlation, they are explained in detail in the above referenced issues. But for fun, lets look at them:

Wrong match due to length:
length

No match:
notile

Wrong match due to wrong tile:
tile_edge

Wrong match due to snapping was patched earlier by changing the tolerance way low. This has now been fixed for good so that even with a higher `NODE_SNAP` tolerance will not snap to the wrong edge because it doesn't search nodes at all.

Ok so now, what about the numbers... I did a regular grid that covers NYC to test performance. This turned out to be a pretty good test case because its a place where we have a lot of data and binning is really important to cut down the search space. Here are the results:

Before binning:

```
2016/03/22 19:28:09.414095 [INFO] Succeeded Searches on Uncached Tiles
2016/03/22 19:28:09.414202 [INFO] --------------------------------
2016/03/22 19:28:09.414214 [INFO] Total: 3025
2016/03/22 19:28:09.414226 [INFO] Fastest: 40.836174,-73.854210 @ 80ms
2016/03/22 19:28:09.414236 [INFO] Slowest: 40.805405,-73.907494 @ 576ms
2016/03/22 19:28:09.414245 [INFO] Median: 291.624463ms
2016/03/22 19:28:09.414253 [INFO] Standard Deviation: 70.768582ms
2016/03/22 19:28:09.414260 [INFO] Faster Than 1 Standard Deviation: 552
2016/03/22 19:28:09.414268 [INFO] Slower Than 1 Standard Deviation: 490
2016/03/22 19:28:09.414275 [INFO] --------------------------------

2016/03/22 19:28:09.414446 [INFO] Failed Searches on Uncached Tiles
2016/03/22 19:28:09.414454 [INFO] --------------------------------
2016/03/22 19:28:09.414461 [INFO] No results
2016/03/22 19:28:09.414468 [INFO] --------------------------------

2016/03/22 19:28:09.414643 [INFO] Succeeded Searches on Cached Tiles
2016/03/22 19:28:09.414651 [INFO] --------------------------------
2016/03/22 19:28:09.414658 [INFO] Total: 3025
2016/03/22 19:28:09.414668 [INFO] Fastest: 40.718227,-73.987411 @ 0ms
2016/03/22 19:28:09.414677 [INFO] Slowest: 40.759251,-73.756531 @ 303ms
2016/03/22 19:28:09.414685 [INFO] Median: 128.940165ms
2016/03/22 19:28:09.414693 [INFO] Standard Deviation: 35.000487ms
2016/03/22 19:28:09.414700 [INFO] Faster Than 1 Standard Deviation: 475
2016/03/22 19:28:09.414707 [INFO] Slower Than 1 Standard Deviation: 468
2016/03/22 19:28:09.414714 [INFO] --------------------------------

2016/03/22 19:28:09.414875 [INFO] Failed Searches on Cached Tiles
2016/03/22 19:28:09.414883 [INFO] --------------------------------
2016/03/22 19:28:09.414889 [INFO] No results
2016/03/22 19:28:09.414896 [INFO] --------------------------------
```

After binning:

```
2016/03/22 19:20:47.819924 [INFO] Succeeded Searches on Uncached Tiles
2016/03/22 19:20:47.820025 [INFO] --------------------------------
2016/03/22 19:20:47.820036 [INFO] Total: 3025
2016/03/22 19:20:47.820045 [INFO] Fastest: 40.825916,-74.040688 @ 5ms
2016/03/22 19:20:47.820052 [INFO] Slowest: 40.718227,-73.987411 @ 497ms
2016/03/22 19:20:47.820058 [INFO] Median: 138.710413ms
2016/03/22 19:20:47.820064 [INFO] Standard Deviation: 107.776862ms
2016/03/22 19:20:47.820069 [INFO] Faster Than 1 Standard Deviation: 423
2016/03/22 19:20:47.820074 [INFO] Slower Than 1 Standard Deviation: 431
2016/03/22 19:20:47.820078 [INFO] --------------------------------

2016/03/22 19:20:47.820271 [INFO] Failed Searches on Uncached Tiles
2016/03/22 19:20:47.820278 [INFO] --------------------------------
2016/03/22 19:20:47.820293 [INFO] No results
2016/03/22 19:20:47.820297 [INFO] --------------------------------

2016/03/22 19:20:47.820542 [INFO] Succeeded Searches on Cached Tiles
2016/03/22 19:20:47.820549 [INFO] --------------------------------
2016/03/22 19:20:47.820564 [INFO] Total: 3025
2016/03/22 19:20:47.820570 [INFO] Fastest: 40.584896,-74.014053 @ 0ms
2016/03/22 19:20:47.820577 [INFO] Slowest: 40.800278,-73.898613 @ 34ms
2016/03/22 19:20:47.820582 [INFO] Median: 4.361983ms
2016/03/22 19:20:47.820587 [INFO] Standard Deviation: 2.982629ms
2016/03/22 19:20:47.820592 [INFO] Faster Than 1 Standard Deviation: 404
2016/03/22 19:20:47.820598 [INFO] Slower Than 1 Standard Deviation: 392
2016/03/22 19:20:47.820602 [INFO] --------------------------------

2016/03/22 19:20:47.820789 [INFO] Failed Searches on Cached Tiles
2016/03/22 19:20:47.820795 [INFO] --------------------------------
2016/03/22 19:20:47.820810 [INFO] No results
2016/03/22 19:20:47.820814 [INFO] --------------------------------
```

Take aways on performance is that:
* tiles with transit information in them are very slow to load and they dominate the time in the uncached scenarios
* binned edges are amazingly fast, in the avg case they are almost as fast as the old school node searching code

If you are interested in the before and afters load this into geojson.io (blue is input, red is old, green is new): 
[diffs.geojson.txt](https://github.com/valhalla/loki/files/184827/diffs.geojson.txt)

### CODE REVIEW

**Comment by dnesbitt61** [comment]
:shipit:

File: test/search.cc [hunk 1]
```diff
@@ -15,48 +15,39 @@
 using namespace valhalla::midgard;
```
Comment (kevinkreiser) [inline test/search.cc]: this file is a bit of a :christmas_tree: but the only changes are, removal of node search and adding of the bins to the test tile which is commented out

**Comment by kevinkreiser** [comment]
closes #62

**Comment by kevinkreiser** [comment]
closes #52

**Pull request merged**